cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,447 @@
+  }
+
+  /**
+   * Update goal state after one round of self-healing / rebalance.
+   *
+   *  @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    ensureBrokersAllHaveEnoughLeaderOfTopics(clusterModel, optimizationOptions);
+    finish();
+  }
+
+  private void ensureBrokersAllHaveEnoughLeaderOfTopics(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+      return; // Early termination to avoid some unnecessary computation
+    }
+    for (Broker broker : clusterModel.aliveBrokers()) {
+      if (!isEligibleToHaveLeaders(broker, optimizationOptions)) {
+        continue;
+      }
+      for (String mustHaveLeaderPerBrokerTopicName : _mustHaveTopicLeadersPerBroker) {
+        int leaderCount = broker.numLeadersFor(mustHaveLeaderPerBrokerTopicName);
+        if (leaderCount < minTopicLeadersPerBroker()) {
+          throw new OptimizationFailureException(String.format("Broker %d does not have enough leader replica for topic %s. "
+              + "Minimum required per-broker leader replica count %d. Actual broker leader replica count %d",
+              broker.id(), mustHaveLeaderPerBrokerTopicName, minTopicLeadersPerBroker(), leaderCount));
+        }
+      }
+    }
+  }
+
+  /**
+   * Rebalance the given broker without violating the constraints of the current goal and optimized goals.
+   * @param broker         Broker to be balanced.
+   * @param clusterModel   The state of the cluster.
+   * @param optimizedGoals Optimized goals.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void rebalanceForBroker(Broker broker,
+                                    ClusterModel clusterModel,
+                                    Set<Goal> optimizedGoals,
+                                    OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    LOG.debug("balancing broker {}, optimized goals = {}", broker, optimizedGoals);
+    moveAwayOfflineReplicas(broker, clusterModel, optimizedGoals, optimizationOptions);
+    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+      return; // Early termination to avoid some unnecessary computation
+    }
+    if (!isEligibleToHaveLeaders(broker, optimizationOptions)) {
```

Comment:
Can we move `broker.isAlive()` out of `isEligibleToHaveLeaders` -- once we apply other proposed changes in this review, it would be redundant to have this check within the method.

```
if (!(broker.isAlive() && isEligibleToHaveLeaders(broker, optimizationOptions))) {
```